### PR TITLE
Authenticated json rpc endpoint

### LINKED ISSUE #4919: Authenticated JSON-RPC endpoint
### Describe the feature

TL;DR:
It would be nice to have an endpoint that exposes exactly what's behind JSON-RPC (8545) but have it under authentication instead.

Full story:
I would like to expose nodes with JSON-RPC (8545) endpoint available but under authentication.
I initially thought leveraging the built-in auth RPC (8551) would be the way to do it as it offers JWT token auth out of the box. Obviously I can setup a reverse proxy that would handle authentication, but ideally I would prefer to avoid that.
However as pointed out by @rkrasiuk and @mattsse on Telegram isn't a good fit because not only auth RPC exposes only a subset of of `eth_` methods (so for instance it doesn't offer `eth_getBalance`), but also it exposes `engine_` methods which we don't want to mess up with.
See spec: https://github.com/ethereum/execution-apis/blob/v1.0.0-beta.3/src/engine/common.md#underlying-protocol
Later we discussed about the possibility to add custom auth support that would guard the existing RPC endpoint.
This would definitely address my requirement so it would be a nice addition.

### Additional context

The Telegram discussion started here https://t.me/paradigm_reth/15965

### PR DESCRIPTION
**Pull request by 1010adigupta**
Closes #4919 
This pr adds JWT authentication to JSON-RPC endpoint, that is it exposes what's behind JSON-RPC (8545) but under authentication

- [x] add cli argument for authentication of http/ws
- [x] add optional JwtSecret field to RpcServerConfig
- [x] add new variant for auth layer in WsHttpServerKind

### CODE REVIEW

File: crates/rpc/rpc-builder/src/lib.rs [hunk 1]
```diff
@@ -1070,7 +1070,7 @@ where
 /// Http and WS share the same settings: [`ServerBuilder`].
 ///
 /// Once the [RpcModule] is built via [RpcModuleBuilder] the servers can be started, See also
-/// [ServerBuilder::build] and [Server::start](jsonrpsee::server::Server::start).
+/// [ServerBuilder::build] and [Server::start](jsonrpc::server::Server::start).
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:1073]: please revert, this is the wrong package

File: crates/rpc/rpc-builder/src/lib.rs [hunk 2]
```diff
@@ -1615,6 +1629,8 @@ enum WsHttpServerKind {
     Plain(Server<Identity, RpcServerMetrics>),
     /// Http server with cors
     WithCors(Server<Stack<CorsLayer, Identity>, RpcServerMetrics>),
+    /// Http server with auth
+    WithAuth(Server<Stack<AuthLayer<JwtAuthValidator>, Identity>, RpcServerMetrics>),
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:1633]: do we also want cors+auth?

File: crates/rpc/rpc-builder/src/lib.rs [hunk 3]
```diff
@@ -1648,6 +1666,20 @@ impl WsHttpServerKind {
             let local_addr = server.local_addr()?;
             let server = WsHttpServerKind::WithCors(server);
             Ok((server, local_addr))
+        } else if let Some(secret) = auth_secret {
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:1669]: this is now mutually exclusive with cors where cors takes precedence over jwt, which is probably not what we want here?

do we want to stack both or should jwt overule cors? 

I think we can just add another variant that stacks cors+auth?

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc-builder/src/lib.rs (hunk 3 above)
Comment (1010adigupta) [reply crates/rpc/rpc-builder/src/lib.rs:1669]: Will add a fourth variant for dealing with cors+auth

File: crates/rpc/rpc-builder/src/lib.rs (hunk 1 above)
Comment (1010adigupta) [reply crates/rpc/rpc-builder/src/lib.rs:1073]: Okay

File: crates/rpc/rpc-builder/src/lib.rs (hunk 3 above)
Comment (1010adigupta) [reply crates/rpc/rpc-builder/src/lib.rs:1669]: or make jwt take precednce over cors?

**Comment by mattsse** [comment]
@1010adigupta let's wrap this up first

**Review by mattsse — APPROVED** [review]
lgtm

I'll wire up the CLI next

**Pull request merged**